### PR TITLE
support requeue header on nack

### PR DESCRIPTION
**Pull request by gm-ghanover**
as of rabbitmq 3.4, the NACK command supports the requeue header, necessary for DLX

### CODE REVIEW

File: src/Protocol/Protocol.php [hunk 1]
```diff
@@ -212,15 +212,19 @@ public function getAckFrame(Frame $frame, $transactionId = null)
      *
      * @param \Stomp\Transport\Frame $frame
      * @param string $transactionId
+     * @param bool $requeue
```
Comment (staabm) [inline src/Protocol/Protocol.php]: please describe what this parameter is used for

**Comment by staabm** [comment]
DLX ?

**Comment by jmglsn** [comment]
The parameter needs to be part of the other protocol versions too, see failing tests here: https://travis-ci.org/stomp-php/stomp-php/jobs/298826443#L1335.

@staabm We need to check what should happen if this is set on the other protocols, they don't provide this functionality.
I would like to see an `LogicException` then. 
In that case the default value must be `null` as `true` would not work.

**Comment by staabm** [comment]
> We need to check what should happen if this is set on the other protocols, they don't provide this functionality.

in case the concept of "requeue" does only exist in certain protcols (or in certain msgqueues) we maybe should use a options array or a parameter object instead of a regular parameter.

File: src/Protocol/Protocol.php (hunk 1 above)
Comment (gm-ghanover) [reply src/Protocol/Protocol.php]: the requeue header on NACK will either requeue the message (default), or discard the message if false. if DLX is enabled for the queue, requeue=false will send the message to the deadletter exchange. see https://www.rabbitmq.com/dlx.html

**Comment by jmglsn** [comment]
@staabm yes, we're ending up in many very specific parameters here. I think for the current version it would be ok just to add a new one. Next major release should reduce this again, I would like to use a parameter object then - it feels like a bigger change and might also reduce the parameters on other protocol functions.
For now I would add `LogicException` for any protocol that is not capable to cover this parameter. This makes sure that nobody uses the parameter and faces unreported data loss.
@gm-ghanover Thanks for checking this, would you like to add the exception for the other protocol implementations? Otherwise we could append this to your PR.

**Comment by gm-ghanover** [comment]
Older versions of rabbitmq (and i assume activemq) that do not support the requeue header will go with their default behavior and requeue the message on NACK (unless otherwise configured at the service level), so I don't think there is any risk of data loss. The only issue that might arise is someone trying to use requeue=false on something which is not supported, and causing confusion when the message is not discarded (or sent to deadletter).

Throwing a LogicException on activemq could make sense to prevent that, but as far as I'm aware, there is no method for determining the version of rabbitmq that is in use, so there is no method to know if the header is supported or not.

File: src/Protocol/Protocol.php [hunk 2]
```diff
@@ -212,15 +212,19 @@ public function getAckFrame(Frame $frame, $transactionId = null)
...
      * @throws StompException
      */
-    public function getNackFrame(Frame $frame, $transactionId = null)
+    public function getNackFrame(Frame $frame, $transactionId = null, $requeue = null)
     {
         if ($this->version === Version::VERSION_1_0) {
             throw new StompException('Stomp Version 1.0 has no support for NACK Frames.');
         }
         $nack = $this->createFrame('NACK');
+        if ($requeue === false) {
```
Comment (jmglsn) [inline src/Protocol/Protocol.php]: It should be possible to send `requeue=>true` too, when I read http://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.nack.requeue it feels as if there is an way to configure the server to change the default behavior, not sure if this applies to the stomp plugin http://www.rabbitmq.com/stomp.html too.
But from my point of view it makes sense to allow true and false to be transmitted.
So the condition should be based on `$requeue !== null`.

File: src/Protocol/Protocol.php (hunk 1 above)
Comment (jmglsn) [reply src/Protocol/Protocol.php]: Please add a small note to the parameter doc, something like "Parameter are not available on all brokers".

File: src/StatefulStomp.php [hunk 3]
```diff
@@ -65,11 +65,12 @@ public function ack(Frame $frame)
      * Not acknowledge consumption of a message from a subscription
      *
      * @param Frame $frame
+     * @param bool $requeue
      * @return void
      */
-    public function nack(Frame $frame)
+    public function nack(Frame $frame, $requeue = null)
```
Comment (jmglsn) [inline src/StatefulStomp.php]: Please add the comment from `Protocol` here too.

File: src/Broker/ActiveMq/ActiveMq.php [hunk 4]
```diff
@@ -95,8 +95,13 @@ public function getAckFrame(Frame $frame, $transactionId = null)
     /**
      * @inheritdoc
      */
-    public function getNackFrame(Frame $frame, $transactionId = null)
+    public function getNackFrame(Frame $frame, $transactionId = null, $requeue = null)
     {
+        if ($requeue !== null) {
+            throw new \LogicException(
+                'requeue header not supported by ActiveMQ. Please read ActiveMQ DLQ documentation.'
```
Comment (staabm) [inline src/Broker/ActiveMq/ActiveMq.php]: why do you mention the docs here? is ActiveMQ DLQ a euqivalent feature as $requeue?

File: src/States/IStateful.php [hunk 5]
```diff
@@ -31,9 +31,10 @@ public function ack(Frame $frame);
      * Not acknowledge consumption of a message from a subscription
      *
      * @param Frame $frame
+     * @param bool $requeue
```
Comment (staabm) [inline src/States/IStateful.php]: here same doc, as requested above.

File: src/Broker/ActiveMq/ActiveMq.php (hunk 4 above)
Comment (gm-ghanover) [reply src/Broker/ActiveMq/ActiveMq.php]: yes, though configured through activemq.xml, not something you can do individually with message headers.

**Comment by gm-ghanover** [comment]
i'll see what i can do about unit tests this weekend, but i don't have docker set up, so you may beat me to it.

**Comment by jmglsn** [comment]
Please squash your commits :)

**Comment by staabm** [comment]
> Please squash your commits :)

@jmglsn  in case you dont know. you can squash via github web ui

grafik

File: src/Protocol/Protocol.php [hunk 6]
```diff
@@ -212,15 +212,19 @@ public function getAckFrame(Frame $frame, $transactionId = null)
      *
      * @param \Stomp\Transport\Frame $frame
      * @param string $transactionId
+     * @param bool $requeue Requeue header supported on RabbitMQ >= 3.4
```
Comment (staabm) [inline src/Protocol/Protocol.php]: shouldn't we implement it in the RabbitMq class then?

Comment (gm-ghanover) [reply src/Protocol/Protocol.php]: that would probably make more sense. should the base protocol then also throw the LogicException if $requeue !== null?

Comment (staabm) [reply src/Protocol/Protocol.php]: yes, I think so.

**Comment by jmglsn** [comment]
@staabm Was not sure about who is going to be author of those commits then, do you know that?

**Comment by staabm** [comment]
@jmglsn authorship will stay intact (not sure what happens in case you have commits of different users though).

I know it works when all commits are from the same user.

**Comment by jmglsn** [comment]
@gm-ghanover Thank you! Your change is included in https://github.com/stomp-php/stomp-php/releases/tag/4.3.1

**Pull request merged**